pcsx2/IPC.cpp

```diff
@@ -0,0 +1,315 @@
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <thread>
+#include <sys/types.h>
+#if _WIN32
+#define read_portable(a, b, c) (recv(a, b, c, 0))
+#define write_portable(a, b, c) (send(a, b, c, 0))
+#define close_portable(a) (closesocket(a))
+#define bzero(b, len) (memset((b), '\0', (len)), (void)0)
+#include <windows.h>
+#else
+#define read_portable(a, b, c) (read(a, b, c))
+#define write_portable(a, b, c) (write(a, b, c))
+#define close_portable(a) (close(a))
+#include <sys/socket.h>
+#include <sys/un.h>
+#endif
+
+#include "Common.h"
+#include "Memory.h"
+#include "System/SysThreads.h"
+#include "IPC.h"
+
+SocketIPC::SocketIPC(SysCoreThread* vm)
+	: pxThread("IPC_Socket")
+{
+#ifdef _WIN32
+	WSADATA wsa;
+	SOCKET new_socket;
+	struct sockaddr_in server, client;
+	int c;
+
+
+	if (WSAStartup(MAKEWORD(2, 2), &wsa) != 0)
+	{
+		Console.WriteLn(Color_Red, "IPC: Cannot initialize winsock! Shutting down...");
+		return;
+	}
+
+	if ((m_sock = socket(AF_INET, SOCK_STREAM, 0)) == INVALID_SOCKET)
+	{
+		Console.WriteLn(Color_Red, "IPC: Cannot open socket! Shutting down...");
+		return;
+	}
+
+	// yes very good windows s/sun/sin/g sure is fine
+	server.sin_family = AF_INET;
+	// localhost only
+	server.sin_addr.s_addr = inet_addr("127.0.0.1");
+	server.sin_port = htons(PORT);
+
+	if (bind(m_sock, (struct sockaddr*)&server, sizeof(server)) == SOCKET_ERROR)
+	{
+		Console.WriteLn(Color_Red, "IPC: Error while binding to socket! Shutting down...");
+		return;
+	}
+
+#else
+	struct sockaddr_un server;
+
+	m_sock = socket(AF_UNIX, SOCK_STREAM, 0);
+	if (m_sock < 0)
+	{
+		Console.WriteLn(Color_Red, "IPC: Cannot open socket! Shutting down...");
+		return;
+	}
+	server.sun_family = AF_UNIX;
+	strcpy(server.sun_path, SOCKET_NAME);
+
+	// we unlink the socket so that when releasing this thread the socket gets
+	// freed even if we didn't close correctly the loop
+	unlink(SOCKET_NAME);
+	if (bind(m_sock, (struct sockaddr*)&server, sizeof(struct sockaddr_un)))
+	{
+		Console.WriteLn(Color_Red, "IPC: Error while binding to socket! Shutting down...");
+		return;
+	}
+#endif
+
+	// maximum queue of 4096 commands before refusing, approximated to the
+	// nearest legal value. We do not use SOMAXCONN as windows have this idea
+	// that a "reasonable" value is 5, which is not.
+	listen(m_sock, 4096);
+
+	// we save a handle of the main vm object
+	m_vm = vm;
+
+	// we start the thread
+	Start();
+}
+
+void SocketIPC::ExecuteTaskInThread()
+{
+	int msgsock = 0;
+	m_end = false;
+
+	// we allocate once buffers to not have to do mallocs for each IPC
+	// request, as malloc is expansive when we optimize for µs.
+	m_ret_buffer = new char[MAX_IPC_RETURN_SIZE];
+	m_ipc_buffer = new char[MAX_IPC_SIZE];
+	while (true)
+	{
+		msgsock = accept(m_sock, 0, 0);
+		if (msgsock == -1)
+		{
+			if ((m_error_count > MAX_SOCKET_ERRORS) || m_end)
+				break;
+			else
+				m_error_count++;
+		}
+		else
+		{
+			if (read_portable(msgsock, m_ipc_buffer, MAX_IPC_SIZE) < 0)
```

Comment:
Is there any guarantee in unix sockets / tcp that says that the message won't be split into multiple pieces?  It seems at least somewhat likely for things from localhost, but I'd rather not have things mysteriously fail because we didn't take this into account